src/main/java/org/aksw/simba/squirrel/analyzer/mime/MimeTypeDetector.java

```diff
@@ -2,36 +2,42 @@
 
 import org.aksw.simba.squirrel.analyzer.impl.RDFAnalyzer;
 import org.apache.jena.riot.Lang;
+import org.apache.jena.riot.RDFLanguages;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.io.File;
 import java.io.FileInputStream;
 import java.io.IOException;
+import java.util.LinkedList;
 
 public class MimeTypeDetector implements TypeDetector {
-    private static final Logger LOGGER = LoggerFactory.getLogger(RDFAnalyzer.class);
+	private static final Logger LOGGER = LoggerFactory.getLogger(RDFAnalyzer.class);
 
     public Lang detectMimeType(File data) {
 
+        LinkedList<FiniteStateMachine> machinesList = new LinkedList<FiniteStateMachine>(); //3.i
         Lang mimeType = null;
+
         try {
             FileInputStream inputStream = new FileInputStream(data.getAbsolutePath());
-            FiniteStateMachine machine1 = RdfAutomata.buildRDFStateMachine();
-            FiniteStateMachine machine2 = TurtleAutomata.buildTurtleStateMachine();
+
+            machinesList.add(FiniteStateMachineFactory.create(RDFLanguages.RDFXML));
+            machinesList.add(FiniteStateMachineFactory.create(RDFLanguages.TURTLE));
+
             char current;
-            while (inputStream.available() > 0 && (!machine1.canStop() || !machine2.canStop())) {
+
+            while (inputStream.available() > 0 && machinesList.size() > 1) { //3.iv (until only one machine is in list)
                 current = (char) inputStream.read();
-                if (!machine1.isError()) //whichever leads to error state is processed anymore.
-                    machine1 = machine1.switchState(String.valueOf(current));
-                if (!machine2.isError())
-                    machine2 = machine2.switchState(String.valueOf(current));
+
+                FiniteStateMachine machine = machinesList.remove(); // (machines gets removed from list for processing)
+                machine.switchState(String.valueOf(current)); //3.ii
```

Comment:
There is an inner loop missing. Let's say there are two machines A and B in the list. The input stream gives you the characters 1, 2 and 3. In that case, your code would show character 1 to A, 2 to B and 3 to A. That is not correct :wink:

Reply:
Initially, both the RDFXML and TURTLE finite state machines are built irrespective of the file. Imagine we are expecting TURTLE as our MIME type. So initially, the file content does get compared with RDFXML state machine since its the first machine built and it reaches the error state after that the file content is compared with TURTLE machine.

Reply:
That is not a good approach. Imagine, you have 1GB XML file. You RDFXML state machine would process the complete file (!) before the TURTLE machine gets the file and fails after the first line.

In my proposed workflow, both machines are executed at the same time. The TURTLE machine will still fail after the first line but the RDFXML machine does not have to process the rest of the file since it is the only machine left which can handle the content. So the result should be RDFXML, right?

Reply:
yes the result should be RDFXML.
So, as suggested by you I have created a ListIterator in which the first character from the inputstream(file) is given to all finite state machines and it will be iterated until one machine is left which will satisfy the rules set respective to the input character and MIME type.